engine/resources/docker_container.go

```diff
@@ -0,0 +1,441 @@
+	ContainerRemoved = "removed"
+
+	// validateCtxTimeout is the length of time, in seconds, before requests
+	// are cancelled in Validate.
+	validateCtxTimeout = 20
+	// checkApplyCtxTimeout is the length of time, in seconds, before requests
+	// are cancelled in CheckApply.
+	checkApplyCtxTimeout = 120
+)
+
+func init() {
+	engine.RegisterResource("docker:container", func() engine.Res { return &DockerContainerRes{} })
+}
+
+// DockerContainerRes is a docker container resource.
+type DockerContainerRes struct {
+	traits.Base // add the base methods without re-implementation
+	traits.Edgeable
+
+	State      string                     `yaml:"state"`      // running, stopped, or removed
+	Image      string                     `yaml:"image"`      // docker image or image:tag
+	Cmd        []string                   `yaml:"cmd"`        // command(s) to run on the container
+	Env        []string                   `yaml:"env"`        // environment variables, e.g. ["VAR"="val",],
+	Ports      map[string]map[int64]int64 `yaml:"ports"`      // port bindings, e.g. {"tcp" => {80 => 8080},},
+	APIVersion string                     `yaml:"apiversion"` // optional api version, e.g. "v1.39"
+
+	Force bool `yaml:"force"` // force destroy and redeploy if image is incorrect
+
+	client *client.Client // docker api client
+
+	init *engine.Init
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *DockerContainerRes) Default() engine.Res {
+	return &DockerContainerRes{}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *DockerContainerRes) Validate() error {
+	ctx, cancel := context.WithTimeout(context.Background(), validateCtxTimeout*time.Second)
+	defer cancel()
+
+	// validate state
+	if obj.State != ContainerRunning && obj.State != ContainerStopped && obj.State != ContainerRemoved {
+		return fmt.Errorf("state must be running, stopped or removed")
+	}
+
+	// TODO: should this stuff happen in validate or do we make sure there's no
+	// whitespace, and let it fail in CheckApply?
+	//
+	// set up temporary docker client to validate image
+	c, err := client.NewClient(client.DefaultDockerHost, obj.APIVersion, nil, nil)
+	if err != nil {
+		return fmt.Errorf("error creating docker client: %s", err)
+	}
+	defer c.Close()
+
+	// validate image
+	resp, err := c.ImageSearch(ctx, obj.Image, types.ImageSearchOptions{Limit: 1})
+	if err != nil {
+		return fmt.Errorf("error searching for image: %s", err)
+	}
+	if len(resp) == 0 {
+		return fmt.Errorf("image: %s not found", obj.Image)
+	}
+
+	// validate env
+	for _, env := range obj.Env {
+		if !strings.Contains(env, "=") || strings.Contains(env, " ") {
+			return fmt.Errorf("invalid environment variable: %s", env)
+		}
+	}
+
+	// validate ports
+	for k, v := range obj.Ports {
+		if k != "tcp" && k != "udp" && k != "sctp" {
+			return fmt.Errorf("ports primary key should be tcp, udp or sctp")
+		}
+		for p, q := range v {
+			if (p < 1 || p > 0xFFFF) || (q < 1 || q > 0xFFFF) {
```

Comment:
why the 0x instead of just an int?

Reply:
I think 0xFFFF better communicates the reason for the check (max value of uint16). Seems less arbitrary than 65536. I wish go had a simple build-in max and min-value methods on number types. How would you prefer I do it, int const?

Reply:
yeah, just put the raw int there. 65535 I think, right? not 65536 ??